src/workload/mod.rs

```diff
@@ -0,0 +1,604 @@
+                    eprintln!("verb: {:?} does not support truncate", command.verb());
+                    std::process::exit(2);
+                }
+            }
+
+            if command.verb().needs_inner_key()
+                && (keyspace.inner_keys_nkeys().is_none() || keyspace.inner_keys_klen().is_none())
+            {
+                eprintln!(
+                    "verb: {:?} requires that `inner_key_klen` and `inner_key_nkeys` are set",
+                    command.verb()
+                );
+                std::process::exit(2);
+            }
+        }
+
+        let command_dist = WeightedAliasIndex::new(command_weights).unwrap();
+
+        Self {
+            keys,
+            key_dist,
+            commands,
+            command_dist,
+            inner_keys,
+            inner_key_dist,
+            vlen: keyspace.vlen().unwrap_or(0),
+            vkind: keyspace.vkind(),
+        }
+    }
+
+    pub fn sample(&self, rng: &mut dyn RngCore) -> Arc<[u8]> {
+        let index = self.key_dist.sample(rng);
+        self.keys[index].clone()
+    }
+
+    pub fn sample_inner(&self, rng: &mut dyn RngCore) -> Arc<[u8]> {
+        let index = self.inner_key_dist.sample(rng);
+        self.inner_keys[index].clone()
+    }
+
+    pub fn gen_value(&self, rng: &mut dyn RngCore) -> Vec<u8> {
+        match self.vkind {
+            ValueKind::I64 => format!("{}", rng.gen::<i64>()).into_bytes(),
+            ValueKind::Bytes => {
+                let mut buf = vec![0_u8; self.vlen];
+                rng.fill(&mut buf[0..self.vlen]);
+                buf
+            }
+        }
+    }
+}
+
+pub async fn reconnect(work_sender: Sender<ClientWorkItem>, config: Config) -> Result<()> {
+    if config.client().is_none() {
+        return Ok(());
+    }
+
+    let rate = config.client().unwrap().reconnect_rate();
+
+    let mut ratelimit_params = if rate.is_some() {
+        Some(convert_ratelimit(rate.unwrap()))
+    } else {
+        // NOTE: we treat reconnect differently and don't generate any
+        // reconnects if there is no ratelimit specified. (In contrast to
+        // request generation where no ratelimit means unlimited).
+        return Ok(());
+    };
+
+    while RUNNING.load(Ordering::Relaxed) {
+        let quanta = if let Some((quanta, ref mut interval)) = ratelimit_params {
+            interval.tick().await;
+            quanta
+        } else {
+            1
+        };
+
+        for _ in 0..quanta {
+            let _ = work_sender.send(ClientWorkItem::Reconnect).await;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn convert_ratelimit(rate: NonZeroU64) -> (u64, Interval) {
+    let rate = u64::from(rate);
+
+    // TODO: this gives approximate rates
+    //
+    // timer granulularity should be millisecond level on most platforms
+    // for higher rates, we can insert multiple work items every interval
+    let (quanta, interval) = if rate <= 1000 {
+        (1, 1000 / rate)
```

Comment:
Shall we worry about the rounding error?
Related- what happened to the rate limiter crate?

Reply:
Still need to circle-back on the ratelimiter rewrite, but we can use the current version here. Replaced in the latest diff.